File: scripts/lib/doctor.py
#!/usr/bin/env python3
# Copyright (C) 2026 Nigel Breslaw
# SPDX-License-Identifier: GPL-3.0-or-later

"""Report local development readiness without contacting a MiSTer."""

from __future__ import annotations

import argparse
import json
import os
from pathlib import Path
import platform
import re
import shutil
import subprocess
import sys
from typing import Any


SCHEMA = "mister-magik-doctor-v1"
SCOPES = ("full-host", "desktop", "arm", "docs", "device")


def command_output(command: list[str]) -> str | None:
    try:
        result = subprocess.run(
            command,
            check=False,
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            text=True,
            timeout=5,
        )
    except (OSError, subprocess.TimeoutExpired):
        return None
    return result.stdout.strip() if result.returncode == 0 else None


class Report:
    def __init__(self, scope: str) -> None:
        self.scope = scope
        self.checks: list[dict[str, Any]] = []

    def add(
        self,
        check_id: str,
        ok: bool,
        detail: str,
        remediation: str | None = None,
        *,
        required: bool = True,
        warning: bool = False,
    ) -> None:
        status = "pass" if ok else ("warn" if warning else "fail")
        self.checks.append(
            {
                "id": check_id,
                "status": status,
                "required": required,
                "detail": detail,
                "remediation": remediation if not ok else None,
            }
        )

    @property
    def ok(self) -> bool:
        return not any(
            check["required"] and check["status"] == "fail"
            for check in self.checks
        )

    def value(self) -> dict[str, Any]:
        return {
            "schema": SCHEMA,
            "scope": self.scope,
            "ok": self.ok,
            "requires_device": self.scope == "device",
            "device_probe": "not_attempted",
            "checks": self.checks,
        }


def require_command(report: Report, name: str, remediation: str) -> bool:
    path = shutil.which(name)
    report.add(
        f"command-{name}",
        path is not None,
        path or f"{name} not found",
        remediation,
    )
    return path is not None


def check_rust_toolchain(root: Path, report: Report, *, require_arm: bool) -> None:
    toolchain_path = root / "apps/mister/rust-toolchain.toml"
    text = toolchain_path.read_text(encoding="utf-8") if toolchain_path.is_file() else ""
    match = re.search(r'^channel = "([^"]+)"$', text, re.MULTILINE)
    channel = match.group(1) if match else None
    if not channel:
        report.add(
            "rust-toolchain",
            False,
            "pinned Rust channel is unreadable",
            "restore apps/mister/rust-toolchain.toml",
        )
        return
    installed = command_output(["rustup", "toolchain", "list"]) or ""
    report.add(
        "rust-toolchain",
        any(line.split("-", 1)[0] == channel or line.startswith(channel) for line in installed.splitlines()),
        f"pinned channel {channel}",
        f"run: rustup toolchain install {channel} --component rustfmt --component clippy",
    )
    components = command_output(
        ["rustup", "component", "list", "--toolchain", channel]
    ) or ""
    for component in ("rustfmt", "clippy"):
        installed_component = any(
            line.startswith(f"{component}-") and line.endswith("(installed)")
            for line in components.splitlines()
        )
        report.add(
            f"rust-component-{component}",
            installed_component,
            f"{component} for {channel}",
            f"run: rustup component add {component} --toolchain {channel}",
        )
    if require_arm:
        targets = command_output(
            ["rustup", "target", "list", "--toolchain", channel]
        ) or ""
        arm_target = "armv7-unknown-linux-gnueabihf"
        report.add(
            "rust-target-armv7",
            any(
                line.startswith(arm_target) and line.endswith("(installed)")
                for line in targets.splitlines()
            ),
            f"{arm_target} for {channel}",
            f"run: rustup target add {arm_target} --toolchain {channel}",
        )


def check_desktop(root: Path, report: Report) -> None:
    for name in ("github-app", "material-icon-theme"):
        path = root / "apps/desktop/vendor" / name
        report.add(
            f"desktop-submodule-{name}",
            (path / ".git").exists(),
            str(path),
            "run: git submodule update --init apps/desktop/vendor/github-app apps/desktop/vendor/material-icon-theme",
        )


def check_docs(root: Path, report: Report) -> None:
    node = command_output(["node", "--version"])
    major = None
    if node:
        match = re.match(r"v(\d+)", node)
        major = int(match.group(1)) if match else None
    report.add(
        "node-version",
        major is not None and major >= 22,
        node or "Node.js unavailable",
        "install Node.js 22 or newer with Corepack",
    )
    pnpm = command_output(
        ["env", "COREPACK_ENABLE_NETWORK=0", "corepack", "pnpm", "--version"]
    )
    report.add(
        "pnpm-version",
        pnpm == "11.10.0",
        pnpm or "Corepack pnpm unavailable",
        "run: corepack prepare pnpm@11.10.0 --activate",
    )
    modules = root / "documentation/node_modules"
    ready = (modules / ".pnpm").is_dir() and os.access(modules / ".bin/astro", os.X_OK)
    report.add(
        "documentation-dependencies",
        ready,
        str(modules),
        "run: pnpm --dir documentation install --frozen-lockfile",
    )


def check_output(root: Path, report: Report) -> None:
    outputs = ("build", "dist", "outputs", "target", "documentation/dist")
    build = root / "build"
    writable = os.access(build if build.exists() else root, os.W_OK)
    report.add(
        "build-output-writable",
        writable,
        str(build),
        "make the repository build output writable",
    )
    for output in outputs:
        sentinel = f"{output}/.mister-magik-doctor"
        ignored = subprocess.run(
            ["git", "-C", str(root), "check-ignore", "-q", sentinel],
            check=False,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        ).returncode == 0
        report.add(
            f"output-ignored-{output.replace('/', '-')}",
            ignored,
            sentinel,
            f"add {output}/ to the repository ignore policy",
        )


def build_report(root: Path, scope: str) -> Report:
    report = Report(scope)
    require_command(report, "python3", "install Python 3")
    require_command(report, "git", "install Git")

    if scope in {"full-host", "desktop", "arm"}:
        require_command(report, "cargo", "install Rust through rustup")
        require_command(report, "rustup", "install rustup")
        check_rust_toolchain(root, report, require_arm=scope == "arm")
    if scope in {"full-host", "desktop"}:
        check_desktop(root, report)
    if scope in {"full-host", "docs"}:
        require_command(report, "node", "install Node.js 22 or newer")
        require_command(report, "corepack", "install Node.js with Corepack")
        check_docs(root, report)
    if scope == "arm":
        if platform.system() == "Darwin" and platform.machine() == "arm64":
            require_command(report, "container", "install and start Apple container")
        else:
            require_command(report, "cross", "install cross 0.2.5")
            require_command(report, "docker", "install and start Docker")
    if scope == "device":
        mister = root / "scripts/mister"
        report.add(
            "device-wrapper",
            os.access(mister, os.X_OK),
            str(mister),
            "restore the executable scripts/mister wrapper",
        )
        report.add(
            "device-network",
            False,
            "network probe intentionally not attempted",
            required=False,
            warning=True,
        )
    entrypoints = ["scripts/validate", "scripts/dev-rust"]
    if scope == "full-host":
        entrypoints.append("scripts/test-host-tools.sh")
    for entrypoint in entrypoints:
        path = root / entrypoint
        report.add(
            f"entrypoint-{entrypoint.replace('/', '-')}",
            os.access(path, os.X_OK),
            str(path),
            f"restore executable permissions on {entrypoint}",
        )
    check_output(root, report)
    return report


def print_human(value: dict[str, Any]) -> None:
    for check in value["checks"]:
        label = check["status"].upper()
        print(f"{label:4} {check['id']}: {check['detail']}")
        if check["remediation"]:
            print(f"     {check['remediation']}")
    print(f"doctor: {'ready' if value['ok'] else 'not ready'} ({value['scope']})")


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--scope", choices=SCOPES, default="full-host")
    parser.add_argument("--json", action="store_true")
    parser.add_argument("--root", type=Path, default=Path(__file__).resolve().parents[2], help=argparse.SUPPRESS)
    args = parser.parse_args()
    value = build_report(args.root.resolve(), args.scope).value()
    if args.json:
        print(json.dumps(value, sort_keys=True))
    else:
        print_human(value)
    return 0 if value["ok"] else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: agent-cli/src/doctor.rs
// Copyright (C) 2026 Nigel Breslaw
// SPDX-License-Identifier: GPL-3.0-or-later

use crate::error::{AgentError, AgentResult};
use crate::model::Outcome;
use crate::progress::{EventKind, Reporter};
use std::fs;
use std::io::Read;
use std::os::unix::ffi::OsStrExt;
use std::os::unix::fs::PermissionsExt;
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::time::Duration;

pub fn execute(repository: &Path, reporter: &mut Reporter<'_>) -> AgentResult<Outcome> {
    let mut failures = Vec::new();
    reporter.emit(
        EventKind::Progress,
        "doctor",
        "Checking host commands",
        Some(10),
    )?;
    for command in ["git", "cargo", "rustup", "node", "corepack"] {
        require_command(command, &mut failures);
    }

    reporter.emit(
        EventKind::Progress,
        "doctor",
        "Checking Rust tooling",
        Some(30),
    )?;
    check_rust(repository, reporter, &mut failures)?;

    reporter.emit(
        EventKind::Progress,
        "doctor",
        "Checking desktop and docs",
        Some(60),
    )?;
    check_desktop(repository, &mut failures);
    check_docs(repository, reporter, &mut failures)?;

    reporter.emit(
        EventKind::Progress,
        "doctor",
        "Checking repository outputs",
        Some(85),
    )?;
    check_outputs(repository, reporter, &mut failures)?;
    let entrypoint = repository.join("scripts/agent");
    if !is_executable(&entrypoint) {
        failures.push("scripts/agent is missing or not executable".into());
    }

    if failures.is_empty() {
        reporter.emit(EventKind::Completed, "doctor", "Host is ready", Some(100))?;
        Ok(Outcome::Passed)
    } else {
        Err(AgentError::Classified {
            code: "host_not_ready",
            detail: failures.join("; "),
        })
    }
}

fn require_command(name: &str, failures: &mut Vec<String>) {
    if find_command(name).is_none() {
        failures.push(format!("{name} is not installed or not on PATH"));
    }
}

fn find_command(name: &str) -> Option<PathBuf> {
    std::env::var_os("PATH")?
        .as_encoded_bytes()
        .split(|byte| *byte == b':')
        .find_map(|directory| {
            let path = Path::new(std::ffi::OsStr::from_bytes(directory)).join(name);
            is_executable(&path).then_some(path)
        })
}

fn is_executable(path: &Path) -> bool {
    fs::metadata(path)
        .is_ok_and(|metadata| metadata.is_file() && metadata.permissions().mode() & 0o111 != 0)
}

fn check_rust(
    repository: &Path,
    reporter: &mut Reporter<'_>,
    failures: &mut Vec<String>,
) -> AgentResult<()> {
    let toolchain = fs::read_to_string(repository.join("apps/mister/rust-toolchain.toml"))
        .map_err(|error| format!("cannot read pinned Rust toolchain: {error}"))?;
    let channel = quoted_value(&toolchain, "channel")
        .ok_or("apps/mister/rust-toolchain.toml has no channel")?;
    let installed = output(repository, reporter, "rustup", &["toolchain", "list"])?;
    if !installed.lines().any(|line| line.starts_with(&channel)) {
        failures.push(format!("Rust toolchain {channel} is not installed"));
    }
    let components = output(
        repository,
        reporter,
        "rustup",
        &["component", "list", "--toolchain", &channel],
    )?;
    for component in ["rustfmt", "clippy"] {
        if !components
            .lines()
            .any(|line| line.starts_with(component) && line.ends_with("(installed)"))
        {
            failures.push(format!(
                "Rust component {component} is missing for {channel}"
            ));
        }
    }
    Ok(())
}

fn check_desktop(repository: &Path, failures: &mut Vec<String>) {
    for name in ["github-app", "material-icon-theme"] {
        if !repository
            .join("apps/desktop/vendor")
            .join(name)
            .join(".git")
            .exists()
        {
            failures.push(format!("desktop submodule {name} is not initialized"));
        }
    }
}

fn check_docs(
    repository: &Path,
    reporter: &mut Reporter<'_>,
    failures: &mut Vec<String>,
) -> AgentResult<()> {
    let node = output(repository, reporter, "node", &["--version"])?;
    let major = node
        .trim_start_matches('v')
        .split('.')
        .next()
        .and_then(|v| v.parse::<u32>().ok());
    if major.is_none_or(|major| major < 22) {
        failures.push("Node.js 22 or newer is required".into());
    }
    let pnpm = output(repository, reporter, "corepack", &["pnpm", "--version"])?;
    if pnpm.trim() != "11.10.0" {
        failures.push("Corepack pnpm 11.10.0 is required".into());
    }
    let modules = repository.join("documentation/node_modules");
    if !modules.join(".pnpm").is_dir() || !is_executable(&modules.join(".bin/astro")) {
        failures.push("documentation dependencies are not installed".into());
    }
    Ok(())
}

fn check_outputs(
    repository: &Path,
    reporter: &mut Reporter<'_>,
    failures: &mut Vec<String>,
) -> AgentResult<()> {
    for output_path in ["build", "dist", "outputs", "target", "documentation/dist"] {
        let sentinel = format!("{output_path}/.mister-magik-doctor");
        let status = command_status(
            repository,
            reporter,
            "git",
            &["check-ignore", "-q", &sentinel],
        )?;
        if !status {
            failures.push(format!("{output_path} is not ignored by Git"));
        }
    }
    Ok(())
}

fn output(
    repository: &Path,
    reporter: &mut Reporter<'_>,
    program: &str,
    args: &[&str],
) -> AgentResult<String> {
    let mut child = Command::new(program)
        .args(args)
        .current_dir(repository)
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .spawn()
        .map_err(|error| AgentError::Classified {
            code: "command_missing",
            detail: format!("cannot start {program}: {error}"),
        })?;
    let status = crate::process::wait(
        &mut child,
        Some(Duration::from_secs(30)),
        program,
        Some(Duration::from_secs(10)),
        || {
            reporter.emit(
                EventKind::Progress,
                "doctor",
                "Inspecting host prerequisites",
                None,
            )
        },
    )?;
    let mut text = String::new();
    child
        .stdout
        .take()
        .ok_or("command stdout unavailable")?
        .read_to_string(&mut text)
        .map_err(|error| format!("cannot read {program} output: {error}"))?;
    if status.success() {
        Ok(text.trim().to_owned())
    } else {
        Err(AgentError::Classified {
            code: "host_probe_failed",
            detail: format!("{program} {} exited with {status}", args.join(" ")),
        })
    }
}

fn command_status(
    repository: &Path,
    reporter: &mut Reporter<'_>,
    program: &str,
    args: &[&str],
) -> AgentResult<bool> {
    let mut child = Command::new(program)
        .args(args)
        .current_dir(repository)
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn()
        .map_err(|error| format!("cannot start {program}: {error}"))?;
    Ok(crate::process::wait(
        &mut child,
        Some(Duration::from_secs(5)),
        program,
        Some(Duration::from_secs(10)),
        || {
            reporter.emit(
                EventKind::Progress,
                "doctor",
                "Inspecting host prerequisites",
                None,
            )
        },
    )?
    .success())
}

fn quoted_value(text: &str, key: &str) -> Option<String> {
    text.lines().find_map(|line| {
        let value = line
            .trim()
            .strip_prefix(key)?
            .trim_start()
            .strip_prefix('=')?
            .trim();
        value
            .strip_prefix('"')?
            .strip_suffix('"')
            .map(str::to_owned)
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::os::unix::fs::PermissionsExt;

    #[test]
    fn parses_pinned_channel() {
        assert_eq!(
            quoted_value("[toolchain]\nchannel = \"1.97.0\"\n", "channel").as_deref(),
            Some("1.97.0")
        );
    }

    #[test]
    fn rejects_unquoted_channel() {
        assert_eq!(quoted_value("channel = stable", "channel"), None);
    }

    #[test]
    fn quoted_values_require_exact_keys_and_balanced_quotes() {
        let text = "channel_name = \"wrong\"\n channel = \"1.97.1\"\nprofile = \"minimal\"\n";
        assert_eq!(quoted_value(text, "channel").as_deref(), Some("1.97.1"));
        assert_eq!(quoted_value(text, "profile").as_deref(), Some("minimal"));
        assert_eq!(quoted_value("channel = \"unterminated", "channel"), None);
        assert_eq!(quoted_value(text, "missing"), None);
    }

    #[test]
    fn executable_and_desktop_prerequisites_are_reported_from_fixtures() {
        let root = std::env::temp_dir().join(format!("agent-doctor-{}", std::process::id()));
        let _ = fs::remove_dir_all(&root);
        fs::create_dir_all(&root).unwrap();
        let command = root.join("tool");
        fs::write(&command, b"#!/bin/sh\n").unwrap();
        assert!(!is_executable(&command));
        let mut permissions = fs::metadata(&command).unwrap().permissions();
        permissions.set_mode(0o755);
        fs::set_permissions(&command, permissions).unwrap();
        assert!(is_executable(&command));

        let mut failures = Vec::new();
        check_desktop(&root, &mut failures);
        assert_eq!(failures.len(), 2);
        for name in ["github-app", "material-icon-theme"] {
            fs::create_dir_all(root.join("apps/desktop/vendor").join(name).join(".git")).unwrap();
        }
        failures.clear();
        check_desktop(&root, &mut failures);
        assert!(failures.is_empty());
        fs::remove_dir_all(root).unwrap();
    }
}
